cc_macro.c

```diff
@@ -779,7 +914,86 @@ struct token_list* maybe_expand(struct token_list* token)
 	{
 		return token->next;
 	}
-	hold2 = insert_tokens(token, hold->expansion);
+
+	struct token_list* expansion = hold->expansion;
+
+	if(token->s[0] == '(')
+	{
+		if(hold->arguments == NULL)
+		{
+			line_error_token(macro_token);
+			fputs("Non-function-like macro '", stderr);
+			fputs(hold->symbol, stderr);
+			fputs("' was called as function-like\n", stderr);
+			exit(EXIT_FAILURE);
+		}
+
+		token = eat_token(token); /* skip '(' */
+
+		expansion = deep_copy_token_list(hold->expansion);
+
+		struct token_list* token_before_macro = token->prev;
+
+		struct token_list* start_token = token;
+		int parens = 1;
+		struct macro_argument* argument = hold->arguments;
+		struct token_list* expand_list;
+		while (parens != 0)
+		{
+			if(token->s[0] == '(')
+			{
+				parens = parens + 1;
+			}
+			else if (token->s[0] == ')')
+			{
+				parens = parens - 1;
+			}
+
+			if ((token->s[0] == ',' && parens == 1) || (token->s[0] == ')' && parens == 0))
+			{
+				token->prev->next = NULL;
+				if(start_token->next != NULL && argument->name != NULL)
+				{
+					line_error_token(start_token);
+					fputs("Function-like macro with arguments of more than one token are not supported.", stderr);
+					exit(EXIT_FAILURE);
+				}
+
+				expand_list = expansion;
+
+				while(expand_list != NULL)
+				{
+					if(match(expand_list->s, argument->name))
+					{
+						expand_list->s = start_token->s;
+					}
+					expand_list = expand_list->next;
+				}
+
+				token->prev->next = token;
+				start_token = token->next;
+				argument = argument->next;
+			}
+
+			token = token->next;
```

Comment:
probably crashy here too...

Reply:
Added a requires that should cover `token->next` and one for `argument->next` at the top.